<a href="https://colab.research.google.com/github/Lalith0024/Section_C_Group_11_Bank_Marketing_Dataset/blob/dev/notebooks/01_extraction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [23]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### Data Extraction: Loading and Initial Inspection of `raw_dataset.csv`

In [24]:
file_path = '/content/raw_dataset.csv'
df = pd.read_csv(file_path, sep=';')
print(f"Successfully loaded data from {file_path}.\n")

Successfully loaded data from /content/raw_dataset.csv.



#### First 5 rows of the DataFrame

In [25]:
display(df.head())

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,56,housemaid,married,basic.4y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
1,57,services,married,high.school,unknown,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
2,37,services,married,high.school,no,yes,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
3,40,admin.,married,basic.6y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
4,56,services,married,high.school,no,no,yes,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no


In [26]:
display(df.tail())

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
41183,73,retired,married,professional.course,no,yes,no,cellular,nov,fri,...,1,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,yes
41184,46,blue-collar,married,professional.course,no,no,no,cellular,nov,fri,...,1,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,no
41185,56,retired,married,university.degree,no,yes,no,cellular,nov,fri,...,2,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,no
41186,44,technician,married,professional.course,no,no,no,cellular,nov,fri,...,1,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,yes
41187,74,retired,married,professional.course,no,yes,no,cellular,nov,fri,...,3,999,1,failure,-1.1,94.767,-50.8,1.028,4963.6,no


#### Dataset Shape

In [27]:
print(f"Dataset has {df.shape[0]} rows and {df.shape[1]} columns.")

Dataset has 41188 rows and 21 columns.


#### Column Names

In [28]:
print(df.columns.tolist())

['age', 'job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays', 'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed', 'y']


#### DataFrame Info: Column data types and non-null counts

In [29]:
display(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

None

#### Descriptive Statistics of numerical columns

In [30]:
display(df.describe())
display(df.describe(include='all'))

,age,duration,campaign,pdays,previous,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed
count,41188.00000,41188.000000,41188.000000,41188.000000,41188.000000,41188.000000,41188.000000,41188.000000,41188.000000,41188.000000
mean,40.02406,258.285010,2.567593,962.475454,0.172963,0.081886,93.575664,-40.502600,3.621291,5167.035911
std,10.42125,259.279249,2.770014,186.910907,0.494901,1.570960,0.578840,4.628198,1.734447,72.251528
min,17.00000,0.000000,1.000000,0.000000,0.000000,-3.400000,92.201000,-50.800000,0.634000,4963.600000
25%,32.00000,102.000000,1.000000,999.000000,0.000000,-1.800000,93.075000,-42.700000,1.344000,5099.100000
50%,38.00000,180.000000,2.000000,999.000000,0.000000,1.100000,93.749000,-41.800000,4.857000,5191.000000
75%,47.00000,319.000000,3.000000,999.000000,0.000000,1.400000,93.994000,-36.400000,4.961000,5228.100000
max,98.00000,4918.000000,56.000000,999.000000,7.000000,1.400000,94.767000,-26.900000,5.045000,5228.100000


,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
count,41188.00000,41188,41188,41188,41188,41188,41188,41188,41188,41188,...,41188.000000,41188.000000,41188.000000,41188,41188.000000,41188.000000,41188.000000,41188.000000,41188.000000,41188
unique,NaN,12,4,8,3,3,3,2,10,5,...,NaN,NaN,NaN,3,NaN,NaN,NaN,NaN,NaN,2
top,NaN,admin.,married,university.degree,no,yes,no,cellular,may,thu,...,NaN,NaN,NaN,nonexistent,NaN,NaN,NaN,NaN,NaN,no
freq,NaN,10422,24928,12168,32588,21576,33950,26144,13769,8623,...,NaN,NaN,NaN,35563,NaN,NaN,NaN,NaN,NaN,36548
mean,40.02406,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2.567593,962.475454,0.172963,NaN,0.081886,93.575664,-40.502600,3.621291,5167.035911,NaN
std,10.42125,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2.770014,186.910907,0.494901,NaN,1.570960,0.578840,4.628198,1.734447,72.251528,NaN
min,17.00000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,1.000000,0.000000,0.000000,NaN,-3.400000,92.201000,-50.800000,0.634000,4963.600000,NaN
25%,32.00000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,1.000000,999.000000,0.000000,NaN,-1.800000,93.075000,-42.700000,1.344000,5099.100000,NaN
50%,38.00000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2.000000,999.000000,0.000000,NaN,1.100000,93.749000,-41.800000,4.857000,5191.000000,NaN
75%,47.00000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,3.000000,999.000000,0.000000,NaN,1.400000,93.994000,-36.400000,4.961000,5228.100000,NaN


#### Check Missing Values

In [31]:
missing_values = df.isnull().sum()
print("Number of missing values per column:")
display(missing_values[missing_values > 0].sort_values(ascending=False))

Number of missing values per column:


,0


#### Check Duplicates

In [32]:
print(f"Number of duplicate rows: {df.duplicated().sum()}")

Number of duplicate rows: 12


#### Unique Values (Column-wise)

In [33]:
print("Number of unique values per column:")
display(df.nunique().sort_values(ascending=False))

Number of unique values per column:


,0
duration,1544
euribor3m,316
age,78
campaign,42
pdays,27
cons.price.idx,26
cons.conf.idx,26
job,12
nr.employed,11
emp.var.rate,10


#### Unique Value Types (Important for EDA later)

In [34]:
for col in df.select_dtypes(include=['object', 'category']).columns:
    if df[col].nunique() < 50: # Display unique values for columns with less than 50 unique entries
        print(f"\nUnique values for column '{col}':")
        print(df[col].unique())
    else:
        print(f"\nColumn '{col}' has {df[col].nunique()} unique values (too many to display all). Top 10 most frequent values:")
        display(df[col].value_counts().head(10))


Unique values for column 'job':
['housemaid' 'services' 'admin.' 'blue-collar' 'technician' 'retired'
 'management' 'unemployed' 'self-employed' 'unknown' 'entrepreneur'
 'student']

Unique values for column 'marital':
['married' 'single' 'divorced' 'unknown']

Unique values for column 'education':
['basic.4y' 'high.school' 'basic.6y' 'basic.9y' 'professional.course'
 'unknown' 'university.degree' 'illiterate']

Unique values for column 'default':
['no' 'unknown' 'yes']

Unique values for column 'housing':
['no' 'yes' 'unknown']

Unique values for column 'loan':
['no' 'yes' 'unknown']

Unique values for column 'contact':
['telephone' 'cellular']

Unique values for column 'month':
['may' 'jun' 'jul' 'aug' 'oct' 'nov' 'dec' 'mar' 'apr' 'sep']

Unique values for column 'day_of_week':
['mon' 'tue' 'wed' 'thu' 'fri']

Unique values for column 'poutcome':
['nonexistent' 'failure' 'success']

Unique values for column 'y':
['no' 'yes']


#### Basic Data Quality Observations

Based on the initial inspection:
- **Missing Values**: There are no missing values in any of the columns, as indicated by `df.isnull().sum()` returning all zeros.
- **Duplicate Rows**: There are 12 duplicate rows present in the dataset.
- **Data Types**: The data types (`int64`, `float64`, `object`) appear appropriate for the content of each column.
- **Unexpected Categories**: Some categorical columns like `job`, `marital`, `education`, `default`, `housing`, and `loan` contain an 'unknown' category. This will need to be addressed during data cleaning or feature engineering, as it represents missing or unclassified information.
- **Separators**: The dataset was loaded successfully using a semicolon (`;`) as a separator, which was crucial for correct parsing.

### Notebook Conclusion

This initial data extraction and inspection revealed the following:
- **Total Rows and Columns**: The dataset contains 41,188 rows and 21 columns.
- **Missing Values**: There are no missing values identified across any of the columns.
- **Duplicate Rows**: A total of 12 duplicate rows were found in the dataset.
- **Data Types**: The dataset comprises a mix of numerical (int64, float64) and categorical (object) data types, which are appropriate for initial analysis.
- **Unique Values**: Categorical columns generally contain a manageable number of unique values, with `job`, `marital`, `education`, `default`, `housing`, `loan`, `contact`, `month`, `day_of_week`, `poutcome`, and `y` being key categorical features.